ovs/lib/helpers/generic/snapshots.py

```diff
@@ -0,0 +1,352 @@
+        return 'Snapshot for vDisk {0} ({1})'.format(self.vdisk_guid, ', '.join(prop_strings))
+
+
+class Bucket(object):
+    def __init__(self, start, end, retention_policy=None):
+        self.start = start
+        self.end = end
+        self.snapshots = []
+        self.retention_policy = retention_policy
+
+    def is_snapshot_in_interval(self, snapshot):
+        # type: (Snapshot) -> bool
+        return self.start >= snapshot.timestamp > self.end
+
+    def try_add_snapshot(self, snapshot):
+        if self.is_snapshot_in_interval(snapshot):
+            self.snapshots.append(snapshot)
+
+    def get_obsolete_snapshots(self, consistency_first=False, bucket_count=0):
+        # type: (bool, int) -> List[Snapshot]
+        """
+        Retrieve all snapshots which are no longer within this interval
+        :param consistency_first: Consistency of the snapshot is prioritized above the age
+        :type consistency_first: bool
+        :param bucket_count: Number of the bucket in the chain. Used to determine if the current snapshot must be consistent
+        :type bucket_count: int
+        :return: List with Snapshots
+        :rtype: List[Snapshot]
+        """
+        _ = consistency_first
+
+        if self.end:
+            snapshot_to_keep = None
+            if self.retention_policy.consistency_first:
+                # Using + 1 as snapshot provided in the consistency_first_on are > 0
+                if self.retention_policy.consistency_first_on and bucket_count + 1 in self.retention_policy.consistency_first_on:
+                    best = None
+                    for snapshot in self.snapshots:
+                        if best is None:
+                            best = snapshot
+                        # Consistent is better than inconsistent
+                        elif snapshot.consistent and not best.consistent:
+                            best = snapshot
+                        # Newer (larger timestamp) is better than older snapshots
+                        elif snapshot.consistent == best.consistent and snapshot.timestamp > best.timestamp:
+                            best = snapshot
+                    snapshot_to_keep = best
+            if not snapshot_to_keep:
+                # First the oldest snapshot and remove all younger ones
+                oldest = None
+                for snapshot in self.snapshots:
+                    if oldest is None:
+                        oldest = snapshot
+                    # Older (smaller timestamp) is the one we want to keep
+                    elif snapshot.timestamp < oldest.timestamp:
+                        oldest = snapshot
+                snapshot_to_keep = oldest
+            _logger.debug('Elected {} as the snapshot to keep within {}.'.format(snapshot_to_keep, self))
+            obsolete_snapshots = [s for s in self.snapshots if s.timestamp != snapshot_to_keep.timestamp]
```

Comment:
you can just compare snapshots here, instead of checking timestamps, right?
`[s for s in self.snapshots if s != snapshot_to_keep`]